src/op/dialect/nccl/nccl.cc

```diff
@@ -505,6 +505,150 @@ class NCCLBroadcast : public NCCLOpEnv {
+  }
+
+ public:
+  ~NCCLAllToAll() {
+    // Nothing
+  }
+
+  std::string name() const override {
+    return TruncateName(GetUniqueName("raf.op.nccl._all_to_all"));
+  }
+
+  void Execute(const CallValues& cv) override {
+    auto args = cv->args.as<raf::op::schema::AllToAllArgs>();
+    Execute({TupleValue::make(ir::Array<Value>(args->x.begin(), args->x.end()))}, cv->out);
+  }
+
+  void Execute(const std::vector<Value>& inputs, value::Value output) {
+    auto comm_ref = GetRef<Communicator>(reinterpret_cast<CommunicatorObj*>(communicator));
+    ncclComm_t nccl_comm = Downcast<NCCLCommunicator>(comm_ref)->nccl_comm;
+    auto input_x = Downcast<value::TupleValue>(inputs[0]);
+
+    if (input_x->fields.size() == 1) {
+      DLTensor* x = input_x->fields[0];
+      DLTensor* out = output;
+
+      int nccl_num_ranks;
+      NCCL_CALL(ncclCommCount(nccl_comm, &nccl_num_ranks));
+      CHECK_EQ(comm_ref->size, nccl_num_ranks)
+          << "NCCL communicator world size does not match with RAF Communicator.";
+
+      int64_t size = 1;
+      for (int i = 0; i < x->ndim; ++i) {
+        size *= x->shape[i];
+      }
+      CHECK(size % nccl_num_ranks == 0) << "Cannot evenly distribute input tensor to all ranks.";
+      int64_t dtype_size_in_bytes = GetSizeInBytes(x->dtype);
+
+      size_t per_rank_bytes = total_input_size / nccl_num_ranks;
+      size_t size_per_rank = per_rank_bytes / dtype_size_in_bytes;
+
+      char* send_buffer = (char*)x->data;
+      char* recv_buffer = (char*)out->data;
+      if (size != 0) {
+        NCCL_CALL(ncclGroupStart());
+        for (size_t i = 0; i < nccl_num_ranks; i++) {
+          NCCL_CALL(ncclSend(send_buffer + i * per_rank_bytes, size_per_rank, dtype, i, nccl_comm,
+                             (cudaStream_t)stream));
+          NCCL_CALL(ncclRecv(recv_buffer + i * per_rank_bytes, size_per_rank, dtype, i, nccl_comm,
+                             (cudaStream_t)stream));
+        }
+        NCCL_CALL(ncclGroupEnd());
+      }
+    } else {
+      int nccl_num_ranks;
+      NCCL_CALL(ncclCommCount(nccl_comm, &nccl_num_ranks));
+      CHECK_EQ(comm_ref->size, nccl_num_ranks)
+          << "NCCL communicator world size does not match with RAF Communicator.";
+
+      // fuse-reorder tensors into a buffer
+      size_t offset = 0;
+      size_t itvl = total_input_size / nccl_num_ranks;
+      for (int i = 0; i < input_x->fields.size(); ++i) {
+        DLTensor* x = input_x->fields[i];
+        size_t size_per_rank = tuple_sizes[i] / nccl_num_ranks;
+        for (int j = 0; j < nccl_num_ranks; ++j) {
+          void* in = reinterpret_cast<uint8_t*>(in_buffer) + offset + j * itvl;
+          void* x_ = reinterpret_cast<uint8_t*>(x->data) + j * size_per_rank;
+          CUDA_CALL(cudaMemcpyAsync(in, x_, size_per_rank, cudaMemcpyDeviceToDevice,
+                                    (cudaStream_t)stream));
+        }
+        offset += size_per_rank;
```

Comment:
I think the alternative of memory copy is just launching more `send` and `receive`. If we use NCCL group call, is there a situation that memory copy cause more overhead, even though memory copy is assigned to another stream?

Reply:
Indeed. We can try to do some profile and see which is better. I guess it depends on message size etc.

Reply:
If so, could we set another parameter to decide whether to perform memory copy? The default value is beneficial for most cases.